### DISCUSSION TITLE
Some questions about the KDTree.C

### DISCUSSION THREAD

**Discussion by Ethan-xj** [idaholab/moose#22707] [Q&A General] [answered]
[KDTree.C](https://github.com/idaholab/moose/blob/e9e9d4d5035cc9f674057c4cc28885b88d196d5c/framework/src/utils/KDTree.C)
```
void
KDTree::neighborSearch(Point & query_point,
                       unsigned int patch_size,
                       std::vector<std::size_t> & return_index,
                       std::vector<Real> & return_dist_sqr)
{
  return_index.resize(patch_size);

  std::size_t n_result =
      _kd_tree->knnSearch(&query_point(0), patch_size, return_index.data(), return_dist_sqr.data());

  if (n_result == 0)
    mooseError("Unable to find closest node!");

  return_index.resize(n_result);
  return_dist_sqr.resize(n_result);
}

```
I want to ask where can I find some info about the `knnSearch`? I didn't find it anywhere else.
Or can somebody tell me what does the `neighborSearch` do? How?

**Comment by GiudGiud** [comment, marked as answer]
Hello

So a good research to get more information about code is the doxygen
Unfortunately the moose doxygen doesnt have comments, but you can refer to the nanoflann one for the routines called here

https://jlblancoc.github.io/nanoflann/classnanoflann_1_1KDTreeSingleIndexAdaptor.html#ad16aad67165bce42ecdb742bc5536173

neighborSearch searches for the closest neighbors of a queried point

Guillaume

**Reply by GiudGiud** [reply]
for the patch_size closest neighbors*

**Reply by Ethan-xj** [reply]
Oh I got it.  patch_size is the number of the closest neighbors right?
I have another question, is the concept 'neighbors' blockrestrictable? I mean, a queried point's neighbor can be another point in other block or not?

**Reply by Ethan-xj** [reply]
And thank you for telling me about the nanoflann, I haven't browsed this labrary.

**Reply by GiudGiud** [reply]
Yes it is the number of closest neighbors the KDTree search will return. 

The KDTree can be block-restricted by creating it only with the points from the desired block.
